File: finance_engine.js
const fs = require('fs');
const path = require('path');

const DB_FILE = path.join(__dirname, 'finance_db.json');

// Reglas de categorización adaptadas al mercado peruano
const CATEGORY_RULES = {
    "Comida & Restaurantes": ["restaurante", "comida", "cena", "almuerzo", "desayuno", "café", "starbucks", "uber eats", "rappi", "pedidosya", "supermercado", "plaza vea", "wong", "metro", "tottus", "vivanda", "tambo", "mass", "oxxo", "panaderia", "menú", "menu"],
    "Transporte & Viajes": ["uber", "indrive", "yango", "cabify", "gasolina", "combustible", "primax", "repsol", "pecsa", "peaje", "estacionamiento", "vuelo", "latam", "pasaje", "taxi"],
    "Software & Tecnología": ["aws", "google cloud", "github", "chatgpt", "openai", "cursor", "hosting", "dominio", "software", "apple", "microsoft", "adobe"],
    "Marketing & Publicidad": ["meta ads", "facebook ads", "google ads", "tiktok ads", "publicidad", "marketing", "imprenta", "volantes"],
    "Servicios & Bancos": ["luz", "luz del sur", "enel", "agua", "sedapal", "internet", "claro", "movistar", "entel", "bitel", "bcp", "interbank", "bbva", "scotiabank", "yape", "plin"],
    "Ventas e Ingresos": ["pago cliente", "ingreso", "ingresó", "ingrese", "venta", "cobro", "transferencia recibida", "factura cobrada", "honorarios", "yape recibido", "plin recibido", "sueldo", "salario"]
};

function loadDB() {
    if (!fs.existsSync(DB_FILE)) {
        const initialData = { transactions: [] };
        fs.writeFileSync(DB_FILE, JSON.stringify(initialData, null, 2));
        return initialData;
    }
    try {
        const raw = fs.readFileSync(DB_FILE, 'utf8');
        return JSON.parse(raw);
    } catch (e) {
        return { transactions: [] };
    }
}

function saveDB(db) {
    fs.writeFileSync(DB_FILE, JSON.stringify(db, null, 2));
}

function categorizeDescription(description) {
    const descLower = description.toLowerCase();
    for (const [category, keywords] of Object.entries(CATEGORY_RULES)) {
        for (const kw of keywords) {
            if (descLower.includes(kw)) {
                return category;
            }
        }
    }
    return "Varios / Otros";
}

function parseWhatsAppMessage(text) {
    if (!text) return null;
    let textClean = text.trim();
    let textLower = textClean.toLowerCase();

    // Reemplazar expresiones comunes como "1 mil", "2 mil", "5 mil", "10 mil" por números reales
    textLower = textLower.replace(/(\d+)\s*mil\b/gi, (match, p1) => {
        return (parseFloat(p1) * 1000).toString();
    });

    // IGNORAR URLs / Enlaces Web
    if (textLower.startsWith("http://") || textLower.startsWith("https://") || textLower.startsWith("www.") || textLower.includes("localhost:")) {
        return null;
    }

    // IGNORAR COMANDOS DE BORRADO / ELIMINACIÓN
    if (["borrar", "eliminar", "deshacer", "undo", "quitar", "anular"].some(kw => textLower.includes(kw))) {
        return null;
    }

    // Ignorar respuestas formateadas del bot
    if (textLower.includes("balance actual") || textLower.includes("registrado exitosamente") || textLower.includes("resumen financiero") || textLower.includes("desglose de gastos") || textLower.includes("eliminado con éxito")) {
        return null;
    }

    const isIncome = ["ingresé", "ingrese", "ingreso", "ingresó", "cobré", "cobre", "cobro", "recibí", "recibi", "venta", "pago cliente", "yape recibido", "plin recibido", "sueldo", "salario"].some(w => textLower.includes(w));
    const transType = isIncome ? "ingreso" : "gasto";

    const amountMatch = textLower.match(/(?:s\/\.?|\$)?\s*(\d+(?:[.,]\d{1,2})?)/i);
    if (!amountMatch) return null;

    const amount = parseFloat(amountMatch[1].replace(',', '.'));
    if (isNaN(amount) || amount <= 0) return null;

    let currency = "PEN";
    if (textLower.includes("usd") || textLower.includes("dólares") || textLower.includes("dolares")) {
        currency = "USD";
    }

    let category = categorizeDescription(textClean);
    if (isIncome && category === "Varios / Otros") category = "Ventas e Ingresos";

    const dateStr = new Date().toISOString().replace('T', ' ').substring(0, 19);

    return {
        source: "whatsapp",
        type: transType,
        amount: amount,
        currency: currency,
        category: category,
        description: textClean,
        date: dateStr
    };
}

function addTransaction(source, type, amount, currency = 'PEN', category = null, description = '', dateStr = null, rawData = null) {
    const db = loadDB();
    const newTx = {
        id: db.transactions.length + 1,
        source: source || 'manual',
        type: type,
        amount: parseFloat(amount),
        currency: currency || 'PEN',
        category: category || categorizeDescription(description),
        description: description,
        date: dateStr || new Date().toISOString().replace('T', ' ').substring(0, 19),
        raw_data: rawData || null,
        created_at: new Date().toISOString()
    };
    db.transactions.push(newTx);
    saveDB(db);
    return newTx;
}

function deleteTransactionsSmart(criteria = {}) {
    const db = loadDB();
    if (db.transactions.length === 0) return [];

    let targetAmount = criteria.target_amount || criteria;
    const deleteAll = criteria.delete_all_matching || false;
    const deletedList = [];

    if (typeof targetAmount === 'number' && targetAmount > 0) {
        for (let i = db.transactions.length - 1; i >= 0; i--) {
            if (Math.abs(db.transactions[i].amount - targetAmount) < 0.01) {
                const deleted = db.transactions.splice(i, 1)[0];
                deletedList.push(deleted);
                if (!deleteAll) break;
            }
        }
    } else {
        const deleted = db.transactions.pop();
        if (deleted) deletedList.push(deleted);
    }

    if (deletedList.length > 0) saveDB(db);
    return deletedList;
}

function getSummary() {
    const db = loadDB();
    let totalIngresos = 0;
    let totalGastos = 0;
    const categoryBreakdown = {};
    const sourceCounts = {};

    for (const tx of db.transactions) {
        if (tx.type === 'ingreso') {
            totalIngresos += tx.amount;
        } else if (tx.type === 'gasto') {
            totalGastos += tx.amount;
            categoryBreakdown[tx.category] = (categoryBreakdown[tx.category] || 0) + tx.amount;
        }
        sourceCounts[tx.source] = (sourceCounts[tx.source] || 0) + 1;
    }

    return {
        total_ingresos: Math.round(totalIngresos * 100) / 100,
        total_gastos: Math.round(totalGastos * 100) / 100,
        balance: Math.round((totalIngresos - totalGastos) * 100) / 100,
        currency_symbol: "S/",
        category_breakdown: categoryBreakdown,
        source_counts: sourceCounts
    };
}

function getAllTransactions(limit = 100) {
    const db = loadDB();
    return db.transactions.slice().reverse().slice(0, limit);
}

module.exports = {
    parseWhatsAppMessage,
    addTransaction,
    deleteTransactionsSmart,
    getSummary,
    getAllTransactions,
    categorizeDescription
};

File: finance_engine.py
import sqlite3
import re
import os
from datetime import datetime

DB_PATH = os.path.join(os.path.dirname(__file__), "finance.db")

# Diccionario de reglas para categorización automática
CATEGORY_RULES = {
    "Comida & Restaurantes": ["restaurante", "comida", "cena", "almuerzo", "desayuno", "café", "starbucks", "uber eats", "rappi", "supermercado", "mercado", "panaderia"],
    "Transporte & Viajes": ["uber", "didi", "cabify", "gasolina", "combustible", "peaje", "estacionamiento", "vuelo", "hotel", "pasaje", "taxi"],
    "Software & Tecnología": ["aws", "google cloud", "github", "chatgpt", "openai", "cursor", "hosting", "dominio", "software", "apple", "microsoft", "adobe"],
    "Marketing & Publicidad": ["meta ads", "facebook ads", "google ads", "tiktok ads", "publicidad", "marketing", "imprenta", "volantes"],
    "Servicios Básicos": ["luz", "agua", "internet", "gas", "teléfono", "claro", "movistar", "entel", "tigo"],
    "Ventas e Ingresos": ["pago cliente", "ingreso", "venta", "cobro", "transferencia recibida", "factura cobrada", "honorarios"],
    "Varios / Otros": []
}

def get_db():
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

def init_db():
    with get_db() as conn:
        conn.execute("""
            CREATE TABLE IF NOT EXISTS transactions (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                source TEXT NOT NULL,          -- 'whatsapp', 'gmail', 'manual'
                type TEXT NOT NULL,            -- 'gasto', 'ingreso'
                amount REAL NOT NULL,
                currency TEXT DEFAULT 'USD',
                category TEXT NOT NULL,
                description TEXT NOT NULL,
                date TEXT NOT NULL,
                raw_data TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        conn.commit()

def categorize_description(description: str) -> str:
    desc_lower = description.lower()
    for category, keywords in CATEGORY_RULES.items():
        for kw in keywords:
            if kw in desc_lower:
                return category
    return "Varios / Otros"

def parse_whatsapp_message(text: str):
    """
    Parsea mensajes de WhatsApp como:
    - 'Gasté 45 USD en supermercado'
    - 'Gasto 25 en taxi'
    - 'Ingresé 500 por venta de servicio'
    - 'Cobré 1200 dólares de cliente X'
    """
    text_clean = text.strip()
    text_lower = text_clean.lower()

    # Detectar Tipo: Gasto o Ingreso
    is_income = any(w in text_lower for w in ["ingresé", "ingreso", "ingresó", "cobré", "cobro", "recibí", "venta", "pago cliente"])
    trans_type = "ingreso" if is_income else "gasto"

    # Extraer Monto
    # Busca patrones de números como 45, 45.50, $45, 45.00
    amount_match = re.search(r'\$?(\d+([.,]\d{1,2})?)', text_clean)
    if not amount_match:
        return None

    amount_str = amount_match.group(1).replace(',', '.')
    try:
        amount = float(amount_str)
    except ValueError:
        return None

    # Extraer Moneda (por defecto USD)
    currency = "USD"
    if "pen" in text_lower or "soles" in text_lower or "s/" in text_lower:
        currency = "PEN"
    elif "eur" in text_lower or "euros" in text_lower:
        currency = "EUR"
    elif "mxn" in text_lower or "pesos" in text_lower:
        currency = "MXN"

    # Asignar Descripción limpia
    description = text_clean
    category = categorize_description(description)
    if is_income and category == "Varios / Otros":
        category = "Ventas e Ingresos"

    date_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    return {
        "source": "whatsapp",
        "type": trans_type,
        "amount": amount,
        "currency": currency,
        "category": category,
        "description": description,
        "date": date_str
    }

def add_transaction(source, trans_type, amount, currency, category, description, date_str=None, raw_data=None):
    if not date_str:
        date_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    with get_db() as conn:
        cursor = conn.cursor()
        cursor.execute("""
            INSERT INTO transactions (source, type, amount, currency, category, description, date, raw_data)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """, (source, trans_type, amount, currency, category, description, date_str, raw_data or ""))
        conn.commit()
        return cursor.lastrowid

def get_summary():
    with get_db() as conn:
        # Totales por tipo
        row_ingresos = conn.execute("SELECT SUM(amount) as total FROM transactions WHERE type='ingreso'").fetchone()
        row_gastos = conn.execute("SELECT SUM(amount) as total FROM transactions WHERE type='gasto'").fetchone()

        total_ingresos = row_ingresos['total'] or 0.0
        total_gastos = row_gastos['total'] or 0.0
        balance = total_ingresos - total_gastos

        # Gastos por categoría
        cat_rows = conn.execute("""
            SELECT category, SUM(amount) as total 
            FROM transactions 
            WHERE type='gasto' 
            GROUP BY category
            ORDER BY total DESC
        """).fetchall()

        category_breakdown = {row['category']: row['total'] for row in cat_rows}

        # Conteo de transacciones por fuente
        sources_rows = conn.execute("""
            SELECT source, COUNT(*) as count 
            FROM transactions 
            GROUP BY source
        """).fetchall()
        source_counts = {row['source']: row['count'] for row in sources_rows}

        return {
            "total_ingresos": round(total_ingresos, 2),
            "total_gastos": round(total_gastos, 2),
            "balance": round(balance, 2),
            "category_breakdown": category_breakdown,
            "source_counts": source_counts
        }

def get_all_transactions(limit=100):
    with get_db() as conn:
        rows = conn.execute("""
            SELECT id, source, type, amount, currency, category, description, date 
            FROM transactions 
            ORDER BY id DESC 
            LIMIT ?
        """, (limit,)).fetchall()
        return [dict(row) for row in rows]

# Inicializa la base de datos al importar el módulo
init_db()
